src/appleseed/foundation/utility/iesparser.cpp

```diff
@@ -0,0 +1,946 @@
+    {
+        if (str == "1")
+        {
+            *type = IESParser::PhotometricTypeC;
+        }
+        else if (str == "2")
+        {
+            *type = IESParser::PhotometricTypeB;
+        }
+        else if (str == "3")
+        {
+            *type = IESParser::PhotometricTypeA;
+        }
+        else
+        {
+            return false;
+        }
+        return true;
+    }
+
+    // Convert string to units type.
+    bool string_to_units_type(
+        const std::string& str, IESParser::LuminousOpeningShape::UnitsType* type)
+    {
+        if (str == "1")
+        {
+            *type = IESParser::LuminousOpeningShape::Feet;
+        }
+        else if (str == "2")
+        {
+            *type = IESParser::LuminousOpeningShape::Meters;
+        }
+        else
+        {
+            return false;
+        }
+        return true;
+    }
+
+    // Split string and convert it to the vector of values.
+    // Returns false, if one of the tokens cannot be converted 
+    // to the value of specified type.
+    template <typename ValueType>
+    bool string_to_vector(
+        const std::string& str, std::vector<ValueType>* output)
+    {
+        std::vector<std::string> tokens;
+        boost::split(tokens, str, isspace, boost::token_compress_on);
+
+        output->clear();
+        output->reserve(tokens.size());
+
+        for (each< std::vector<std::string> > token = tokens; token; ++token)
+        {
+            try
+            {
+                output->push_back(boost::lexical_cast<ValueType>(*token));
+            }
+            catch (boost::bad_lexical_cast&)
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    template <typename ValueType>
+    bool check_increasing_order(const std::vector<ValueType>& values)
+    {
+        for (int i = 1; i < values.size(); ++i)
+        {
+            if (values[i] <= values[i - 1])
+                return false;
+        }
+        return true;
+    }
+}
+
+//
+// Class implementation.
+//
+
+IESParser::IESParser() :
+    ignore_allowed_keywords(false),
+    ignore_required_keywords(false),
+    ignore_empty_lines(true),
+    ignore_tilt(false)
+{}
+
+
+void IESParser::parse(std::istream& input_stream)
+{
+    reset(input_stream);
+
+    // Parse version.
+    parse_format_version(input_stream);
+
+    // Parse block before TILT and TILT line.
+    parse_keywords_and_tilt(input_stream);
+
+    // Parse TILT data if it is included.
+    if (tilt_specification == IncludeTilt)
+    {
+        parse_tilt_data(input_stream);
+    }
+
+    // Parse the rest of the data.
+    parse_photometric_data(input_stream);
+    parse_angles(input_stream);
+    parse_candela_values(input_stream);
+
+    if (!input_stream.eof())
+    {
+        throw ParsingException("Expected end of file", line_counter);
+    }
+}
+
+
+bool IESParser::is_keyword_allowed_by_iesna02(const std::string& keyword)
+{
+    if (keyword == "TEST" ||
```

Comment:
While this is perfectly fine, a small simplification is to write:

    return
        keyword == "TEST" ||
        keyword == "TESTLAB" ||
        ...